runtime/src/bank.rs

```diff
@@ -1791,6 +1792,21 @@ pub fn goto_end_of_slot(bank: &mut Bank) {
     }
 }
 
+// This protects from memory exhaustion.
+// Because we're are about to restore full process state from file, allow use
+// of memory as much as possible. But set a hard-limit with ample of head rooms.
+// This is an extra protection for possible bincode bugs in addition to
+// deserialize_snapshot_data_file
+pub fn deserialize_for_snapshot<R, T>(reader: R) -> bincode::Result<T>
+where
+    R: Read,
+    T: serde::de::DeserializeOwned,
+{
+    bincode::config()
+        .limit(MAX_SNAPSHOT_DATA_FILE_SIZE)
```

Comment:
Well after some search, this can't be reverted to before completely. We need this restriction. Otherwise, bicode can easily tricked allocating too much of memory. `bincode` can't know the input size in advance at all because bincode is fed by the input merely with `Write`. So, it can't determine sensible limits without out-of-band config like this.

Refs:
- https://docs.rs/bincode/1.2.1/bincode/struct.Config.html#byte-limit-details
- https://github.com/servo/bincode/pull/42/files
- https://github.com/servo/bincode/commit/628b989fdfe855cfed7641f4f1896960b66f19e4